File: 백준/Silver/17626. Four Squares/Four Squares.py
import math

def is_square(k):
    sqrt_k = int(math.isqrt(k))
    return sqrt_k * sqrt_k == k

n = int(input().strip())

# 1개 제곱수로 표현 가능한 경우
if is_square(n):
    print(1)
else:
    # 2개 제곱수로 표현 가능한지 확인
    for i in range(1, int(math.isqrt(n)) + 1):
        if is_square(n - i * i):
            print(2)
            break
    else:
        # 3개 제곱수로 표현 가능한지 확인
        for i in range(1, int(math.isqrt(n)) + 1):
            for j in range(1, int(math.isqrt(n - i * i)) + 1):
                if is_square(n - i * i - j * j):
                    print(3)
                    break
            else:
                continue
            break
        else:
            # 위의 경우에 해당하지 않으면 4
            print(4)

File: 백준/Silver/17626. Four Squares/Four Squares.java
import java.io.*;
import java.util.*;

public class Main {

    // 제곱수인지 확인하는 메서드
    private static boolean isSquare(int k) {
        if (k < 0) return false; // 음수는 제곱수가 아님
        int sqrtK = (int) Math.sqrt(k);
        return sqrtK * sqrtK == k;
    }

    // 제곱수의 합으로 표현 가능한 최소 개수를 찾는 메서드
    private static int findMinSquares(int n) {
        // 1개 제곱수로 표현 가능한 경우
        if (isSquare(n)) return 1;

        // 2개 제곱수로 표현 가능한지 확인
        for (int i = 1; i * i <= n; i++) {
            if (isSquare(n - i * i)) return 2;
        }

        // 3개 제곱수로 표현 가능한지 확인
        for (int i = 1; i * i <= n; i++) {
            for (int j = 1; j * j <= (n - i * i); j++) {
                if (isSquare(n - i * i - j * j)) return 3;
            }
        }

        // 위의 경우에 해당하지 않으면 4
        return 4;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        sc.close();

        // 최소 제곱수 개수를 찾아 출력
        System.out.println(findMinSquares(n));
    }
}
